### ISSUE TITLE
Discuss approaches to take backup of Zeebe without downtime

### ISSUE THREAD

**Issue by deepthidevaki** [camunda/zeebe#9260]

**Comment by deepthidevaki** [comment]
### Challenges when taking hot backup of zeebe

When taking backup of zeebe, the state of of all partitions in the backup should be consistent. I have described the problems and challenges in detail [here](https://docs.google.com/document/d/1dNg2LR6TVEFGSfGIe2gnHEvoOKeef3t5TUm578t5c-s/edit#) and [here](https://docs.google.com/presentation/d/1T5HhFmIXmI_pc3TogCEuQAV0X0Sv-hD8D7xe91e9SsY/edit#slide=id.gee76306af0_0_135). Here is the summary:

**State across partitions**
Consider the following scenario:
![image](https://user-images.githubusercontent.com/1997478/166230899-3dc107b8-95a1-471a-8005-f6c4b3b613b4.png)

Partition 1 takes the checkpoint before deploying version 2 and Partition 2 takes the checkpoint after deploying version 2. When zeebe restores from these checkpoints, Partition 1 has v1 and partition 2 has v2. When we create instances of this process, it will create instances of different version depending on which partition it was created. 

Similar issues can happen with message correlation. One partition thinks that the message was correlated, while the other one is still waiting for the message. 

**State across replicas of same partition**
Inconsistencies can also happen if we checkpoint the replicas of a partition independently. 
![image](https://user-images.githubusercontent.com/1997478/166231753-9edfc8b0-4ffa-4e3f-97a5-0953af608bb1.png)

If zeebe restores from these checkpoints, the properties of raft are violated. Node 0 has committed entries which are not in the other two replicas thus violating the quorum.

**Comment by deepthidevaki** [comment]
There are two ways to address the above challenges.
1. Let the backup have these inconsistencies. But on restore the brokers can identify them and fix them.
2. Coordinate the checkpointing process so that they are consistent. 

I don't know yet if fixing the inconsistencies during restore is feasible. I will discuss some ideas for coordinated checkpoints in the following comments.

**Comment by deepthidevaki** [comment]
### Coordinated backup - Algorithm to take a global checkpoint

One way to take a coordinated checkpoint without blocking all processes is [BCS algorithm as proposed in this paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.149.5565&rep=rep1&type=pdf). (An easy to understand version is [here](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.78.4041&rep=rep1&type=pdf)). This is a communication induced checkpointing and relies on lazy coordination among the participating processes. 

Here is how the algorithm works.

C(i, _sni_) is a local checkpoint of process i, and it checkpoint id is _sni._ When a process sends a message, it also embeds it current checkpoint id _sni_ with it.

A process takes it local checkpoint when
1. A basic checkpoint is triggered. A basic checkpoint is triggered either by an external trigger or it could be something that is triggered periodically.
2. A forced checkpoint is triggered. A forced checkpoint happens when a process received a message from another process and it is forced to take a checkpoint to guarantee the consistency. 

The algorithm is as follows:

```
When a basic checkpoint is scheduled:
		sni = sni + 1
		take checkpoint C(i,sni)
		
Upon the receipt of a message m:
		if sni < m.sn
			then sni = m.sn
				 take checkpoint C(i, sni)

```

A global checkpoint C = set of local checkpoints from all processes.
A global checkpoint C is a [consistent cut](https://docs.google.com/presentation/d/1T5HhFmIXmI_pc3TogCEuQAV0X0Sv-hD8D7xe91e9SsY/edit#slide=id.gee76306af0_0_164) if all local checkpoints in it has the same id. A system can recover from a global checkpoint if it is consistent cut.

**Comment by deepthidevaki** [comment]
# Coordinated backup in Zeebe

Let's see how the above algorithm can be applied to take backups in Zeebe.

### What is in a zeebe backup?

Applying the algorithm in zeebe 
- a process = a partition.
- A message between processes is any command that are send between partitions - such as commands related to deployment, message correlation etc.
- Zeebe backup consists of backups of all partitions. The backup only includes one copy of a partition. That means we only take the backup from the leader, but not from the followers. When restoring, all replicas can restore from the same backup.

**Some notation to make it easy for discussions:**

A command in logstream = C(position, lastCheckpointId)
An event in logstream = (E, position).
An example logStream :[S(10), C(10, 1), E(11), C(12,1), E(13), C(13,2), E(14), C(14,2)...] where S(10) indicates the snapshot at position 10.

A zeebe backup with id = `y`, consists of each partition's checkpoint with id `y`.

The checkpoint `y` of a partition consists of:

Snapshot: S_i (Snapshot at position i)
Log: [record(i) ... record(i+n)] 
 where record at i+n is the first command with lastCheckpointId = y.
 
In the above example logstream. the backup 2 contains the logStream
[S(10), C(10, 1), E(11), C(12,1), E(13), C(13,2)] because C(13,2) is the first command with lastCheckpointId = 2.


### Backup process (high level overview)

1. StreamProcessor knows lastCheckpointId of its partition.
1. All commands send between partitions piggyback the lastCheckpointId of the sender.
1. When a partition process a command
	1. if the command.lastCheckpointId > local lastCheckpointId
		- Updates lastCheckpointId = command.lastCheckpointId
		- Takes the checkpoint `y` of the partition

To start taking a backup:
1. Write a new command `checkpoint` with lastCheckpointId = the next backupId to all partitions.
	- *Note:-* We can also write this command to only one partition. The other partitions will take a checkpoint, when it receives a message from this partition. However, we don't know when a message will be send. So a partition may never take a checkpoint because it never receives a command from another partitions. To prevent this, we trigger a "basic" checkpoint on all partitions by writing a command to all partitions.


### Challenges
The above process looks simple, but there are several caveats to it.

#### Atomicity of taking a backup
**Concurrent snapshots**
When the backup of partition is taken, it should include the snapshot at position < checkpoint position. Since the snapshotting process is concurrent to streamprocessor, we would need some coordination here. Otherwise following scenarios can happen while taking a backup: 
	- A later snapshot is added to the backup
	- The snapshot that should be included in the backup is deleted.

**Concurrent writes to the log**
While taking a backup, raft is concurrently writing new records to the log. It may also have uncommitted records. When we just copy the segment files, it may contain records written after the record `checkpoint-y`. 

A potential solution here is to allow records after `checkpoint-y` in the backup, but on restore we delete all records after `checkpoint-y`.

#### Failures
Reasons for failure:
* Leader change. The backup process started, but  did not complete before a role change.
* Any other errors such i/o failures when copying data to backup storage.

If there are two commands with same checkpointId, then the backup should be taken for the first message. If for any reason, it failed to take backup, it should not take another backup with the same id for the second message. If it takes a second backup with same id at a different position, the backups are no more consistent. If a backup fails, the next backup should be taken with a later id.

Backup process can also take long. So we do not want to block the StreamProcessor until the backup is completed (success or failure).

Note:- In theory two replica's of same partition can take backup for the same `checkpoint-y` record. Semantically they both are valid backups. However in practice, they can write to the same locations thus corrupting each other's backup. So if we allow this, we should take extra precautions.

#### Message reordering
Commands sends between partitions are not through a reliable channel. How does this affect the consistency of the backup. Should we enforce a reliable message delivery?

**Comment by deepthidevaki** [comment]
#### Solution to overcome some of the above challenges

The following actors are involved in the backup process

- StreamProcessor
- BackupActor
- SnapshotDirector
- SnapshotStoreActor

StreamProcessor:
- When a command is read
	1. if command.lastCheckpointId > local lastCheckpointId
		- send request `takeBackup command.lastCheckpointId command.position` to BackupActor
		- update local `lastCheckpointId = command.lastCheckpointId`
			- lastCheckpointId is stored in the database
		- write followup event - `checkpoint-taken lastCheckpointId`
			- During replay, updates lastCheckpointId in database. But never initiates the checkpoint. 
	2. Process the command if it is a normal engine command
- All commands send between the partitions should have the lastCheckpointId of the sender

SnapshotDirector:
Before committing the snapshot check the following:
If `snapshotId.processedPosition < streamProcessor.lastCheckpointId < lastWrittenPosition`, then abort the snapshot. This is needed to prevent the scenario where the concurrently taken snapshot has `processedPosition` > `checkpoint position`.


BackupActor:
- On receiving the command `takeBackup checkpointId position`
	- get the current snapshot from SnapshotStore
	- If `currentSnapshot.processedPosition`  < `checkpointId`
		- then "lock" the snapshot by sending a command to SnapshotStore
		- If successfully locked - continue with the backup
			- Take backup
				- Copy snapshot files
				- Copy Segment files
			- Once backup is completed, unlock the snapshot
		- else mark the backup as failed.
	- else 
		- mark the backup as failed. The latest snapshot is already past the current checkpoint. So we cannot take a valid checkpoint.


SnapshotStoreActor:

- On snapshot lock request
	- Mark the snapshot as locked in memory
	- A locked snapshot should not be deleted until it is unlocked
	- When a snapshot is locked, the corresponding logs should be not compacted as well (TODO: how to do it)
- On snapshot unlock request
	- Remove the in-memory lock
	- Delete the snapshot if there is a newer snapshot
	- Compact the logs
	
On Restore:
Delete all records after the checkpoint record. The backup should only included the snapshot and records until the checkpoint position. checkpoint position = position of the first record with lastCheckpointId = backup id.

**Comment by oleschoenburg** [comment]
Could you expand a little bit on this section:
> SnapshotDirector:
Before committing the snapshot check the following:
If snapshotId.processedPosition < streamProcessor.lastCheckpointId < lastWrittenPosition, then abort the snapshot. This is needed to prevent the scenario where the concurrently taken snapshot has processedPosition > checkpoint position.

The last sentence makes sense to me but I don't understand how that is related to the condition "snapshotId.processedPosition < streamProcessor.lastCheckpointId < lastWrittenPosition".

**Comment by deepthidevaki** [comment]
> Could you expand a little bit on this section:
> 
> > SnapshotDirector:
> > Before committing the snapshot check the following:
> > If snapshotId.processedPosition < streamProcessor.lastCheckpointId < lastWrittenPosition, then abort the snapshot. This is needed to prevent the scenario where the concurrently taken snapshot has processedPosition > checkpoint position.
> 
> The last sentence makes sense to me but I don't understand how that is related to the condition "snapshotId.processedPosition < streamProcessor.lastCheckpointId < lastWrittenPosition".

@oleschoenburg 

Oops. Sorry. There is a typo in there it should be `snapshotId.processedPosition < lastCheckpointPosition < lastWrittenPosition`, where lastCheckpointPosition = position of the record that triggered the backup.

Before taking a snapshot, SnapshotDirector gets `streamProcessor.processedPosition` and this is used as `snapshotId.processedPosition`. Then the snapshotting process is started. StreamProcessor is concurrently processing records, so by the time snapshot is taken it may have processed more records and updated the processed position in the state. If `lastWrittenPosition` is less than the lastCheckpointPosition, then we are sure that the snapshot is not past the lastCheckpointPosition. So we can take a valid backup.

**Comment by Zelldon** [comment]
I guess I need to take some more time to understand in all details how it should work, but I guess it sounds like a solution, but also a "lot" of additional complexity. I guess :thinking: In general the taking backup from the leaders makes total sense to me. This was something I also ask you last week so sounds good to me :+1: 

During reading, I had a thought, which I was not able to drop. _What if we just backup the snapshots?_ Do we need the log at all? 

I think we can restore Zeebe we just a snapshot right? Could we couple the backup and snapshot logic more? Like if a snapshot is taken a backup is taken as well ? OR we allow to trigger snapshots from outside to allow the same logic as you have described with backup commands? Furthermore, for me it sounds like backups could be equal to snapshots (backup == snapshot). 

In the snapshot we have all data we have processed and applied, plus exported. We just need to make sure that this is consistent with the other partitions and dependencies (elastic etc.) 

Might be a bad idea, not sure but would like to clarify this.

**Comment by deepthidevaki** [comment]
> What if we just backup the snapshots? Do we need the log at all?

Our snapshots are not complete without the followup events in the log, right? This is why in SnapshotDirector we wait until the `lastWrittenPosition` is committed before it commits a snapshot. So we cannot restore Zeebe with just a snapshot.

**Comment by oleschoenburg** [comment]
I don't really understand how it is guaranteed that all partitions eventually reach the same checkpoint id. I can see how that works when only considering resource deployment since that involves sending commands to all partitions which then update the checkpoint id and take a forced checkpoint.

But what about message correlation? Since that only involves two partitions (sender & receiver), wouldn't other partitions potentially never increase the checkpoint id and therefor a global checkpoint can't be reached?

**Comment by Zelldon** [comment]
No. On processing a command we apply all events. If we take a snapshot the event changes are inside that. Meaning we can restore (and we do) with the snapshot (without applying the events again) and start from the snapshot position. I think we wait to make sure that all changes have been persisted to the log as well, but not sure whether it is really necessary since we anyway would start from the snapshot position. On replay we check for the lastProcessedPosition, if we find nothing we just do our normal processing.

**Comment by deepthidevaki** [comment]
> No. On processing a command we apply all events. If we take a snapshot the event changes are inside that. Meaning we can restore (and we do) with the snapshot (without applying the events again) and start from the snapshot position. I think we wait to make sure that all changes have been persisted to the log as well, but not sure whether it is really necessary since we anyway would start from the snapshot position. On replay we check for the lastProcessedPosition, if we find nothing we just do our normal processing.

Consider the following scenario:

In Log: `WorkflowInstanceCreation:Create | WorkflowInstanceCreation:Created, Process:Activate_Element`
StreamProcessor processed the `Create` command and writes the follow up event, `created ` and followup command `Activate_Element`.  
We take the snapshot at this point. So processed position = position of `Create` command. If we restore from this snapshot without the log, the workflow instance will be in `created` state in the snapshot. But it will not continue its execution because `Process:Activate_Element` command is not in the log. As far as I know the instance will stay in the `created` state for ever.

**Comment by deepthidevaki** [comment]
> I don't really understand how it is guaranteed that all partitions eventually reach the same checkpoint id (which is necessary to get to a global checkpoint). I can see how that works when only considering resource deployment since that involves sending commands to all partitions which then update the checkpoint id and take a forced checkpoint.
> 
> But what about message correlation? Since that only involves two partitions (sender & receiver), wouldn't other partitions potentially never increase the checkpoint id and therefor a global checkpoint can't be reached?

@oleschoenburg 

If we only rely on messages between partition to trigger the backup, then it is possible that some partition may never take a checkpoint. What we can do is - to write a command  `checkpoint` with the expected checkpointId  to all partitions when a backup is triggered. So all partition will take the checkpoint with same id more or less around same time. In case there are message send between the partitions during this period, the checkpointId in the message will guarantee the consistency.

**Comment by oleschoenburg** [comment]
> to write a command checkpoint with the expected checkpointId

thanks, that was the missing piece :+1: 

> In case there are message send between the partitions during this period, the checkpointId in the message will guarantee the consistency

In this case the backup might fail in the sense that a consistent global checkpoint won't be reached, right? So it would need to be retriggered externally. 

So in a sense, sending the checkpointId along with messages/deploy commands only serves the purpose of (semantically) aborting the backup process. There is no guarantee that once a backup is triggered, it will happen succeed at some point in the future. Instead the backup will need to be re-triggered until we catch just the right moment where no cross-partition communication invalidates the backup.  Do I understand that correctly? Tbh. I had hoped for something slightly different, where once a backup is triggered, the backup algorithm eventually finds that right moment automatically (by continuously communicating & taking new checkpoints) without the need for external triggers.

**Comment by Zelldon** [comment]
> > No. On processing a command we apply all events. If we take a snapshot the event changes are inside that. Meaning we can restore (and we do) with the snapshot (without applying the events again) and start from the snapshot position. I think we wait to make sure that all changes have been persisted to the log as well, but not sure whether it is really necessary since we anyway would start from the snapshot position. On replay we check for the lastProcessedPosition, if we find nothing we just do our normal processing.
> 
> Consider the following scenario:
> 
> In Log: `WorkflowInstanceCreation:Create | WorkflowInstanceCreation:Created, Process:Activate_Element` StreamProcessor processed the `Create` command and writes the follow up event, `created ` and followup command `Activate_Element`. We take the snapshot at this point. So processed position = position of `Create` command. If we restore from this snapshot without the log, the workflow instance will be in `created` state in the snapshot. But it will not continue its execution because `Process:Activate_Element` command is not in the log. As far as I know the instance will stay in the `created` state for ever.

Thanks @deepthidevaki that is a very good point! I totally forgot about the follow commands, which drive the execution. You're right.

**Comment by Zelldon** [comment]
>  I had hoped for something slightly different, where once a backup is triggered, the backup algorithm eventually finds that right moment automatically (by continuously communicating & taking new checkpoints) without the need for external triggers.

I agree with ole here :)

**Comment by deepthidevaki** [comment]
> In this case the backup might fail in the sense that a consistent global checkpoint won't be reached, right? So it would need to be retriggered externally.
> 
> So in a sense, sending the checkpointId along with messages/deploy commands only serves the purpose of (semantically) aborting the backup process. There is no guarantee that once a backup is triggered, it will succeed at some point in the future. Instead the backup will need to be re-triggered until we catch just the right moment where no cross-partition communication invalidates the backup. Do I understand that correctly?

That is not the case. Receiving a message will force a partition to take a checkpoint, if a checkpoint with same Id doesn't exists. It doesn't have to abort the backup process. 

There are two ways a checkpoint is taken:

Case 1: When a checkpoint is triggered. In Zeebe this would be by writing a special record `checkpoint` to the logstream.

![image](https://user-images.githubusercontent.com/1997478/166902567-589d7a3a-f68b-445e-ae68-45e5dad817ba.png)

Both partitions take the checkpoint when it reads the `checkpoint` command.

Case 2: When a checkpoint is forced by a remote message:

![image](https://user-images.githubusercontent.com/1997478/166902720-da3acf16-9219-4740-be73-cf34f0bb49f2.png)

In this case partition 2 received the message before it recieved the checkpoint command. It takes the checkpoint before the message is processed to guarantee consistency. It then ignores the checkpoint command received later because a checkpoint with the same id already exists. This is ok, because checkpoint that is already taken is consistent with checkpoints of other partition with the same checkpoint id.

Does this help you to understand?

**Comment by Zelldon** [comment]
Just I get it right. Everytime we (the partition) receive a message command or deployment it will create a checkpoint? This allows us to be consistent?

**Comment by oleschoenburg** [comment]
> receive a message command or deployment it will create a checkpoint

Only if that message carries a checkpoint id that is larger than the own local checkpoint id, right?

**Comment by deepthidevaki** [comment]
> Just I get it right. Everytime we (the partition) receive a message command or deployment it will create a checkpoint? This allows us to be consistent?


No. It has to take a checkpoint only if command.checkpointId > last local checkpointId

```
StreamProcessor:

    When a command is read
        if command.lastCheckpointId > local lastCheckpointId
            send request takeBackup command.lastCheckpointId command.position to BackupActor

```

**Comment by Zelldon** [comment]
Cool gotcha! See? We all aligned now :grin: Thanks both :bow: :D

**Comment by Zelldon** [comment]
Just another question regarding the backup itself. Where is this then stored? I hope not on the same disk?

**Comment by deepthidevaki** [comment]
> Just another question regarding the backup itself. Where is this then stored? I hope not on the same disk?

That's the next problem. We are not there yet :smile: 

We could store it anywhere. It should be in a remote storage so that during restore all replicas can restore from the same copy. 

Storing the backup to a remote storage is the responsibility of BackupActor. It can copy the files to anywhere in background without interrupting the StreamProcessor. What we have not discussed is what would happen if it takes too long to copy, or if it fails during this process.

**Comment by oleschoenburg** [comment]
> Does this help you to understand?

Thanks @deepthidevaki, that was just what I needed to understand :+1: 


There is one more thing I'm missing though :see_no_evil: 


What prevents this blue cut:
![drawing-1](https://user-images.githubusercontent.com/1379753/166910372-42bbf7b8-db83-4f46-a256-a4599fcbf783.svg)

If I understand correctly, this would result in a backup with inconsistencies: P1 thinks it has sent a message but P2 knows nothing about the message. 

Maybe this problem doesn't really exist and it just looks like that because the graphics simplify the messaging between partitions and don't include the ACK arrow?

**Comment by deepthidevaki** [comment]
@oleschoenburg 

This is a good point. Also a bit tricky. In short - this is not a problem. 

When StreamProcessor sends the message it also marks it in the state. We have an inbuilt retry mechanism when ack is not received. In CP-1 of partiton1 the message is already in the state, and in CP-1 of partition it has not received or acknowledged. When we recover from this checkpoint, the message will be resend.

Note:- When the application expects a reliable communication channel, it won't retry sending messages. The original algorithm handles this by making the send messages as part of the checkpoint. On recovery it knows which messages to resend (it only defines the theory, it doesn't tell how to do it practically). In Zeebe, we don't expect reliable communication and so retrying is part of the application's (StreamProcessor) duty. This is a slight deviation from the original algorithm, which I think will not affect the correctness.

**Comment by npepinpe** [comment]
I have several questions, but feel free to tell me these go too into implementation and we'll deal with them later :smile: I had Ole's question as well, curious to see the answer. Also, I hope I didn't miss anything, apologies if some of my questions were already answered :see_no_evil: 

1. How is the complete backup orchestrated? So everyone takes locally some checkpoint, but this has to be collected and combined for restoring later.
2. Do I understand correctly that only leaders are taking checkpoints/local backups? 
3. How do we guarantee that the cluster-wide backup process will halt/produce a backup? If I understood correctly, a global backup is consistent only if it all the local checkpoints in it have the same ID. Considering that we're dealing with a distributed system, it's possible that checkpoint messages between partitions are lost, fail, etc.
4. When the backup actor fails, does it need to notify the sender of the original checkpoint (if any)? Do we retry the local backup?
5. How will this scale as more partitions are added to the system? Do we need to expect anything there?
6. What's the expected behavior if I receive several consecutive checkpoint requests at the same time? Could this have a noticeably negative impact on the system as a whole?
7. How will we make this process observable for debugging/diagnosis?

**Comment by deepthidevaki** [comment]
@npepinpe I don't have detail answers for all of your questions. We should discuss them once we are convinced that the above algorithm is correct. But I will try to answer some of them.


> * How is the complete backup orchestrated? So everyone takes locally some checkpoint, but this has to be collected and combined for restoring later.

I assume each partition will write the backup to a shared remote storage. To know if a backup is completed, an external orchestrator can check this remote storage and see if all partitions have "completed" the backup. Or we can query Zeebe to know the backup status. In this case Gateway will be the orchestrator and can collect the backup status from all partitions.

> 
> * Do I understand correctly that only leaders are taking checkpoints/local backups?

Yes. I think it is simple this way. If followers also take snapshot, we would also have to coordinate between the replicas.

> 
> * How do we guarantee that the cluster-wide backup process will halt/produce a backup? If I understood correctly, a global backup is consistent only if it all the local checkpoints in it have the same ID. Considering that we're dealing with a distributed system, it's possible that checkpoint messages between partitions are lost, fail, etc.

The messages between partitions will follow Zeebe's usual messaging pattern. So all messages will be retried until ack is received. So I think it is safe to assume that all partitions will trigger the checkpoint at somepoint. But there are cases where it fails or do not complete - eg:- during the backup there is leader change, or fails due to I/0 exceptions when accessing the remote storage. This is something yet to be discussed. 

> 
> * When the backup actor fails, does it need to notify the sender of the original checkpoint (if any)? Do we retry the local backup?

We haven't discussed the failure scenarios yet. But my current idea is to mark the backup as failed - so we know that a backup exists at that id, but it is invalid and we should not use it for restoring. 

> 
> * How will this scale as more partitions are added to the system? Do we need to expect anything there?

All partitions takes checkpoint independently and in parallel. So more number of partitions should not affect this in anyway. 

> 
> * What's the expected behavior if I receive several consecutive checkpoint requests at the same time? Could this have a noticeably negative impact on the system as a whole?

If there are several checkpoint requests with increasing checkpoint id, then multiple checkpoints will be triggered. This can affect the performance depending on how we implement the BackupActor. If there are multiple checkpoint requests with the same Id, the checkpoint is triggered only once. 

> 
> * How can we make this process observable for debugging/diagnosis?

Haven't thought about it. One idea would be like the ElasticSearch snapshot. There is an api to request the status of the snapshot given the snapshot name. It returns if the snapshot is ongoing, completed or failed. We could do something similar.

**Comment by deepthidevaki** [comment]
@npepinpe Next, I will write down the failure scenarios and how it may affect the backups. So some of your questions can be then answered clearly.

**Comment by deepthidevaki** [comment]
> Is cross-partition communication idempotent? Meaning, when restoring from a backup leads to P1 re-sending a message to P2, is that allowed?

Yes. Duplicate messages can happen in normal operation as well. For example, P1 send message to P2, New leader elected for P1, P2 send ack to old leader, New leader of P1 resend the message after timeout.

**Comment by npepinpe** [comment]
@oleschoenburg Yes and yes :)

I don't think reordering of messages will have any effect on backups, assuming the backup is consistent in itself.

**Comment by Zelldon** [comment]
Since I'm more the visual guy I draw some sequence diagrams.

I think in the **Happy Path it should work**


|       |    |     |
|-----|-----|----|
|![Example1-happypath](https://user-images.githubusercontent.com/2758593/167087813-c2a366c6-fc8a-4fd7-9bd4-b7704492f15e.png)|![Example2-happypath](https://user-images.githubusercontent.com/2758593/167087816-a8d66f8d-5235-4aeb-9e4d-feebc7a72bb1.png)|![Example3-happypath](https://user-images.githubusercontent.com/2758593/167087819-98429bf5-f40f-4c7c-952c-86fac0b83970.png)|

Not sure regarding failure scenarios like retry message or reordering. We have to make sure to send the right last backup ID I guess.

![Example3-happypath (1)](https://user-images.githubusercontent.com/2758593/167087999-4b3dafe8-8f7a-4b8b-90b2-6414e48f6fa7.png)

I our Backup N+1 we would contain the correlation of the first message, but we would also not create a new backup as far as I understood?

**Comment by deepthidevaki** [comment]
> I our Backup N+1 we wouldn't contain the correlation of the first message, but we would also not create a new backup as far as I understood?

@Zelldon Yes. In this case Backup N+1 is also consistent. In the sender the message is in "pending" state, and the receive has not received it. So by the default Zeebe behaviour, the message will be resend. 

Now the question is what should be the backup id when we resend the message. I would say it can use the initial backup id itself. If we abstract away the resending, then we could just consider it as if the message was delayed and re-ordered as follows:

![image](https://user-images.githubusercontent.com/1997478/167093004-72bcdc6d-aa9c-497b-933e-651d8e72c07e.png)

That means, the message is retried with the same backup id. In this case Backup 2 is consistent. 

Consider also another scenario:

![image](https://user-images.githubusercontent.com/1997478/167094415-200dc404-2779-4cac-89bf-242299bc813d.png)

In this case, In Backup-2 msg2 is correlated, but msg1 is not. But this is also ok because when zeebe recovers from Backup2, it will resend msg1.

**Comment by Zelldon** [comment]
Ok regarding the second example. What if you switch Backup 2 with Ack msg-1. This would mean on one of the partition we have in the backup the Ack, but on the other we have not the correlation right?

**Comment by deepthidevaki** [comment]
Like this?
![image](https://user-images.githubusercontent.com/1997478/167096180-6342a9b2-3b84-44d5-a683-d5f089bb96cf.png)

This is similar to a normal failure scenario in zeebe where Ack is lost. The sender will resend msg1, but it will be rejected by the receiver saying it is already correlated. That is the semantics, right?

**Comment by Zelldon** [comment]
No I meant on the above line that the ack is inside the Backup 2 but sending the ack in the line on the bottom is not part of the backup. Does this make sense?

**Comment by deepthidevaki** [comment]
> No I meant on the above line that the ack is inside the Backup 2 but sending the ack in the line on the bottom is not part of the backup. Does this make sense?

Aah Ok.

Ack would also contains the backup id. Because it is also a remote command send by the StreamProcessor. When the ack is received, first a forced backup is taken only then the state is updated. So the backup2 would not contain the acknowledgment.

![image](https://user-images.githubusercontent.com/1997478/167097386-d5c251b7-7f1e-4e03-828e-83d7b3db28d4.png)

**Comment by deepthidevaki** [comment]
**Failure scenarios**
Here are some scenarios where the backup can fail or interrupted before it is completed.

Scenario 1:
1. StreamProcessor sends "takeBackup" command to BackupActor
2. StreamProcess writes "checkpoint taken" record
3. Leader change before "checkpoint taken" is committed
4. BackupActor did not start taking checkpoint before the leaderchange

Scenario 2:
1. StreamProcessor sends "takeBackup" command to BackupActor
2. StreamProcess writes "checkpoint taken" record
3. BackupActor starts taking checkpoint
3. Leader change before "checkpoint taken" is committed


Scenario 3: (this case is very less likely)
1. StreamProcessor sends "takeBackup" command to BackupActor
2. StreamProcess writes "checkpoint taken" record
3. BackupActor starts taking chackpoint
4. BackupActor completes taking the checkpoint
3. Leader change before "checkpoint taken" is committed

Scenario 4:
1. StreamProcessor sends "takeBackup" command to BackupActor
2. StreamProcess writes "checkpoint taken" record
3. BackupActor starts taking snapshot
4. checkpoint taken record is committed
5. Failover happens before backup is completed

Scenario 5:
1. StreamProcessor sends "takeBackup" command to BackupActor
2. StreamProcess writes "checkpoint taken" record
3. BackupActor starts taking snapshot
4. checkpoint taken record is committed
5. Backup failes due to some errors - such as i/o exceptions

Scenario 6:
1. StreamProcessor sends "takeBackup" command to BackupActor
2. StreamProcess writes "checkpoint taken" record
4. checkpoint taken record is committed
5. Failover happens before BackupActor starts taking backup


Scenario 7:
In addition to the above scenarios, the algorithm can also fail a backup when the pre-conditions are not met. For example Backup actor takes a backup only if the following condition is met `currentSnapshot.processedPosition  < checkpointId`. Otherwise the backup should be marked as failed.


For scenario 1,2 and 3, the new leader will process `checkpoint` command and starts taking the backup for the same id. This is ok semantically. But if there is already a (partial) backup in the backup storage, we should ensure that the new leader is not writing to the same location and corrupting the backup. Here we could do something similar to how we handle concurrent snapshots.

Scenario 5 is ok. When there is an error, the leader can mark the backup as failed. We could also retry it a fixed number of times assuming they are temporary errors.But if it fails, it is nothing we can do other than mark the backup as failed.

Scenario 4 is tricky. The new leader will only replay the "checkpoint taken" record, so it won't start taking the backup. The backup will be in a non-completed state for ever.

Scenario 6 is similar to 4, the new leader doesn't start taking the backup. Thus the backup process will remain uncompleted.

Scenario 7 is required to prevent invalid backups. It would be good if we can ensure that such scenarios never happens.

**Comment by Zelldon** [comment]
Ok thanks @deepthidevaki I think this is an important fact to know / to keep in mind **"first a forced backup is taken only then the state is updated."**

**Comment by deepthidevaki** [comment]
> I think this is an important fact to know / to keep in mind **"first a forced backup is taken only then the state is updated."**

Yes this is important.

I realize now that I made a mistake in defining the backup. In the example given here https://github.com/camunda/zeebe/issues/9260#issuecomment-1115827226
The backup 2 should consists of [S(10), C(10, 1), E(11), C(12,1), E(13)].   C(13,2) must not be included in the backup because if it is included then it is similar to taking backup after receiving the message.

**Comment by deepthidevaki** [comment]
There are two more failure scenarios in addition to what is described here https://github.com/camunda/zeebe/issues/9260#issuecomment-1119389831

Scenario 4b:
1. StreamProcessor sends "takeBackup" command to BackupActor
2. StreamProcess writes "checkpoint taken" record
3. BackupActor starts taking snapshot
4. checkpoint taken record is committed
5. New snapshot is taken with processedposition > checkpoint position
6. Failover happens before backup is completed
7. New leader starts with the new snapshot at processedPosition > checkpoint position

Scenario 6b:
1. StreamProcessor sends "takeBackup" command to BackupActor
2. StreamProcess writes "checkpoint taken" record
3. checkpoint taken record is committed
4. New snapshot is taken with processedPosition > checkpoint position
5. Failover happens before BackupActor starts taking backup
6. New leader starts with the new snapshot at processedPosition > checkpoint position

This is similar to scenario 4 and 6. But the difference is that in 4b and 6b, the `checkpoint-taken` records is not replayed as it is already in the snapshot.

**Comment by deepthidevaki** [comment]
**Handling failures**

A failover can result in a failed backup. In most cases, the new leader cannot complete or retake the backup. So it make sense to consider the backup as failed. However, we have to actively mark the backup as failed, otherwise it may remain in ongoing status forever. One way to do it will be as follows: After the failover `BackupActor` of the new leader marks all ongoing backups of that partition as failed.

BackupActor:
- On Reprocessing completed:
        - Gets the lastCheckpointId from StreamProcessor
        - the status of lastCheckpoint by querying remote storage
        - If it is not completed, mark it as failed.

This will cover the scenarios 4,4b,6 and 6b.

To handle scenarios 1,2 and 3, we can allow multiple ongoing backups for a partiton. Similar to multiple ongoing snapshot, we can assign them temporary ids. BackupActor should then atomically commits the ongoing backup to mark it as completed. If there is an ongoing backup from the old leader and one from the new leader, the new leader will commits its ongoing backup.

**Comment by deepthidevaki** [comment]
**Restore Process**
We had previously discussed about concurrent writes to the log while taking the backup. I will summarize it again here. The backup should contain the snapshot and the log until the checkpoint position. However while taking the backup, raft will be writing new entries to the log concurrently. To simplify taking the backup, we can include extra entries in the backup. Instead, during restore we will fix it by deleting all entries that should not be in the backup. So on Restore:

- Find the first command with lastCheckpointId = backupId. Let's call this checkpointCommand.
- Delete all records after checkpointCommand.position including the checkpointCommand.

However in this state, according to StreamProcessor lastCheckpointId = backupId - 1. This is not correct, as it may trigger another backup with the same backupId. To prevent this, we can write a new command `Restore backupId`. StreamProcessor on processing this command can just update the state and set lastcheckpointId = backupId.

**Comment by Zelldon** [comment]
Thanks @deepthidevaki 

I have some questions

> the coordinator resend the command if needed.

How does the coordinator knows? You wrote you will send a response after the command is processed, so at this time the retry mechanism will stop right? So we wouldn't retry failed backups?

> **BackupStore**

Is it a separate remote application?

> Backup[N][p][b].files = contains backup N of partition p started by broker b

I guess this is log and state ?


> A locked snapshot should not be deleted until it is unlocked
> When a snapshot is locked, the corresponding logs should be not compacted as well (TODO: how to do it)

I think this has some potential for race conditions. 


> Note:- Restore process should be completed before any normal operation of Zeebe including raft leader election. No new records should be written to the log before the restore process is completed.

So this needs to be done on bootstrap, right? Alternatively, we could have a separate application, like an init container which does this?



Some parts were not 100% clear to me, I guess because I have not the full picture or the reason behind certain things. For example the checkpointId is now equal to the last process position? 


I think it makes sense to write a ZEP as you mentioned in slack. I'm happy to review this and raise my question then if you think it makes more sense there.

**Comment by deepthidevaki** [comment]
> > the coordinator resend the command if needed.
> 
> How does the coordinator knows? You wrote you will send a response after the command is processed, so at this time the retry mechanism will stop right? So we wouldn't retry failed backups?

If a backup failed, then there is no use in retrying it. But if the request to a partition was lost, it makes sense for the coordinator to resend the request. Receiving duplicate commands on the partition is fine, because it will only trigger backup for the first command. For the second command, StreamProcessor can just send a n ack that to the coordinator.

> > **BackupStore**
> 
> Is it a separate remote application?
 
"BackupStore is a remote storage where each partition can backup it's data" 
It is a remote storage - eg:- Remote file storage, Cloud storage, database etc.. which ever we decide to support as a backup storage.

> > A locked snapshot should not be deleted until it is unlocked
> > When a snapshot is locked, the corresponding logs should be not compacted as well (TODO: how to do it)
> 
> I think this has some potential for race conditions.

Implementation details. I'm pretty sure we can find a way to do it.

> > Note:- Restore process should be completed before any normal operation of Zeebe including raft leader election. No new records should be written to the log before the restore process is completed.
> 
> So this needs to be done on bootstrap, right? Alternatively, we could have a separate application, like an init container which does this?

Implementation details. There might be several ways to achieve this. Using init container seems to be a good option.

> 
> Some parts were not 100% clear to me, I guess because I have not the full picture or the reason behind certain things. For example the checkpointId is now equal to the last process position?

Where do you see that checkpointId is lastProcessedPosition?
StreamProcessor updates its state as follows:
```
update local lastCheckpointId = command.lastCheckpointId
update lastCheckpointPosition = command.position
```

> I think it makes sense to write a ZEP as you mentioned in slack. I'm happy to review this and raise my question then if you think it makes more sense there.

If you don't see any blockers, I will start writing ZEP. I don't want to waste time on writing ZEP if the team is not convinced with this approach.

**Comment by Zelldon** [comment]
On receiving the command takeBackup checkpointId position

get the current snapshot from SnapshotStore
If currentSnapshot.processedPosition < checkpointId

> > > the coordinator resend the command if needed.
> > 
> > 
> > How does the coordinator knows? You wrote you will send a response after the command is processed, so at this time the retry mechanism will stop right? So we wouldn't retry failed backups?
> 
> If a backup failed, then there is no use in retrying it. But if the request to a partition was lost, it makes sense for the coordinator to resend the request. Receiving duplicate commands on the partition is fine, because it will only trigger backup for the first command. For the second command, StreamProcessor can just send a n ack that to the coordinator.

> If a backup failed, then there is no use in retrying it.

Not sure. Why you think so?

> 
> > > **BackupStore**
> > 
> > 
> > Is it a separate remote application?
> 
> "BackupStore is a remote storage where each partition can backup it's data" It is a remote storage - eg:- Remote file storage, Cloud storage, database etc.. which ever we decide to support as a backup storage.
> 
> > > A locked snapshot should not be deleted until it is unlocked
> > > When a snapshot is locked, the corresponding logs should be not compacted as well (TODO: how to do it)
> > 
> > 
> > I think this has some potential for race conditions.
> 
> Implementation details. I'm pretty sure we can find a way to do it.

Not sure whether these are really implementation details. I see this as part of the concept and should be underestimated. It is similar to what you described when to backup which elastic index etc.



> 
> > > Note:- Restore process should be completed before any normal operation of Zeebe including raft leader election. No new records should be written to the log before the restore process is completed.
> > 
> > 
> > So this needs to be done on bootstrap, right? Alternatively, we could have a separate application, like an init container which does this?
> 
> Implementation details. There might be several ways to achieve this. Using init container seems to be a good option.
> 

Might be interesting to have in general an application that is in charge of the complete process. Asking myself whether it is really necessary to interact with zeebe. We could also just get access to the PVC and copy the data :thinking: After copying we check whether everything is valid, checksum etc. and that records from the log are correctly copied. If not we just retry. I'm currently not sure what the benefit is of having it inside of zeebe? :thinking: maybe you can remind me again.

> > Some parts were not 100% clear to me, I guess because I have not the full picture or the reason behind certain things. For example the checkpointId is now equal to the last process position?
> 
> Where do you see that checkpointId is lastProcessedPosition? StreamProcessor updates its state as follows:
> 
> ```
> update local lastCheckpointId = command.lastCheckpointId
> update lastCheckpointPosition = command.position
> ```
> 
> > I think it makes sense to write a ZEP as you mentioned in slack. I'm happy to review this and raise my question then if you think it makes more sense there.
> 
> If you don't see any blockers, I will start writing ZEP. I don't want to waste time on writing ZEP if the team is not convinced with this approach.

Would the ZEP cover only zeebe or everything? Because I think the coordination and architecture is interesting to consider as well. Might be useful to have some images to see the flow and interactions etc.

**Comment by deepthidevaki** [comment]
> On receiving the command takeBackup checkpointId position
> 
> get the current snapshot from SnapshotStore If currentSnapshot.processedPosition < checkpointId
> 

Well, that is a typo :wink: It should be If currentSnapshot.processedPosition < position

> > If a backup failed, then there is no use in retrying it.
> 
> Not sure. Why you think so?

If we re-try with the same id, the back is not triggered again, because the backup should be taken only for the first command with that checkpointId. 

> > Implementation details. I'm pretty sure we can find a way to do it.
> 
> Not sure whether these are really implementation details. I see this as part of the concept and should be underestimated. It is similar to what you described when to backup which elastic index etc.

I think it's too much details at this point, Might make sense to look at it when prototyping how to solve this.

> > > > Note:- Restore process should be completed before any normal operation of Zeebe including raft leader election. No new records should be written to the log before the restore process is completed.
> > > 
> > > 
> > > So this needs to be done on bootstrap, right? Alternatively, we could have a separate application, like an init container which does this?
> > 
> > 
> > Implementation details. There might be several ways to achieve this. Using init container seems to be a good option.
> 
> Might be interesting to have in general an application that is in charge of the complete process. Asking myself whether it is really necessary to interact with zeebe. We could also just get access to the PVC and copy the data thinking After copying we check whether everything is valid, checksum etc. and that records from the log are correctly copied. If not we just retry. I'm currently not sure what the benefit is of having it inside of zeebe? thinking maybe you can remind me again.

I never said it has to be inside zeebe. I said init container seems to be a good option. :thinking: Are you taking about restore or something else?

> 
> Would the ZEP cover only zeebe or everything? Because I think the coordination and architecture is interesting to consider as well. Might be useful to have some images to see the flow and interactions etc.

It make sense to have two documents. One for zeebe which covers this proposal and related concerns within zeebe. Another document for c8 proposals covered in #9222.

**Comment by npepinpe** [comment]
OK, I hope I'm not going to make anyone repeat themselves... :sweat_smile: I read the proposal and have some comments/questions:

- As we store files under Backup[N][p][b], presumably this solves the issue of a leader change occurring during the backup - the new leader would just start putting things under a different separation, e.g. Backup[N][p][b] could translate to `backups/N/p/b/` as folder structure. How would we deal with consecutive leader changes, e.g. 1 -> 2 -> 1? I suppose deleting any previous work would be enough (if possible and assuming we still have a valid snapshot for the checkpoint ID). I think leader changes is going to be a very interesting source of, um, "edge" cases :smile: 
- Who marks the backup completed? For the start, you mentioned that the coordinator sends messages "reliably" (i.e. retries until it receives an acknowledgement) - I didn't see anything related to this on completion, but I assume we would do the same, only reversed? Each leader sends a complete command to the coordinator, retrying until the coordinator acknowledges it? Then the coordinator marks the backup as completed?
- What happens when we fail early because there's no snapshot < position? I see some potential for recovery (i.e. tell the coordinator, and see if we can't take a backup with some new greater ID - though that might send us in a loop).
- Regarding locking the snapshot - I guess we need to disable compaction, then trigger it when it's re-enabled.
- Regarding copying the segments - we probably only want to copy segments up to the segment containing our an entry with checkpoint ID > then the requested one? Do we need to copy all segments? If taking a backup is slow, this might cause us to copy more data than necessary, further slowing things down.
- I assume whenever mark the status of some backup (whether global or local for a partition/broker), this is done via communication with the coordinator?
- As we already have a coordinator, why not use it as a central point for, well, coordination? We can define a thin protocol between the BackupActor/Broker and the BackupCoordinator. StartBackup, FailBackup, CompleteBackup, etc. Possibly the coordinator could handle leader changes as well? Not sure, to be honest. I guess I'm going into implementation details again.
- Having the StreamProcessor process the command to take a backup...as we're talking about an engine interface, is this something we still want to do? I assume it would be a stream processor we (our team) maintain, outside of the engine. Could this be a problem? Do we need to access the state or something like that? From reading into it, it looks like we'd need to fit this concept into the API the automation team would have to implement somewhat. Can we have a clearer separation of concern from our POV? i.e. we deal with logs, and whatever the snapshot store has (which as far as we're concerned is some metadata + a directory of files + SFV for checksum). Having the engine process these commands couples how we take backup with the engine implementation. Ideally we don't need that.

I think the idea is sound in general, but I also feel like it needs to be fleshed out, and a prototype will definitely help us find all these wonderful "edge" cases :smile:

**Comment by deepthidevaki** [comment]
> * As we store files under Backup[N][p][b], presumably this solves the issue of a leader change occurring during the backup - the new leader would just start putting things under a different separation, e.g. Backup[N][p][b] could translate to `backups/N/p/b/` as folder structure. How would we deal with consecutive leader changes, e.g. 1 -> 2 -> 1? I suppose deleting any previous work would be enough (if possible and assuming we still have a valid snapshot for the checkpoint ID). I think leader changes is going to be a very interesting source of, um, "edge" cases smile

I think I covered this case here https://github.com/camunda/zeebe/issues/9260#issuecomment-1120719504

**Comment by deepthidevaki** [comment]
> * Who marks the backup completed? For the start, you mentioned that the coordinator sends messages "reliably" (i.e. retries until it receives an acknowledgement) - I didn't see anything related to this on completion, but I assume we would do the same, only reversed? Each leader sends a complete command to the coordinator, retrying until the coordinator acknowledges it? Then the coordinator marks the backup as completed?

I see multiple ways to do it. 1. The last partition that commits its backup can marks the clusterwide backup as completed. 2. Coordinator can periodically query the status of the partition backup and aggregrate the results. Optionally, add some timeout to mark the backup as failed. 3. What you said.

**Comment by deepthidevaki** [comment]
> * What happens when we fail early because there's no snapshot < position? I see some potential for recovery (i.e. tell the coordinator, and see if we can't take a backup with some new greater ID - though that might send us in a loop).

Then we just mark the backup as failed. Operator can retry backup with a new ID.

**Comment by deepthidevaki** [comment]
> * Regarding copying the segments - we probably only want to copy segments up to the segment containing our an entry with checkpoint ID > then the requested one? Do we need to copy all segments? If taking a backup is slow, this might cause us to copy more data than necessary, further slowing things down.

Have covered it here https://github.com/camunda/zeebe/issues/9260#issuecomment-1120789213
We can think about optimization later.

**Comment by deepthidevaki** [comment]
> * As we already have a coordinator, why not use it as a central point for, well, coordination? We can define a thin protocol between the BackupActor/Broker and the BackupCoordinator. StartBackup, FailBackup, CompleteBackup, etc. Possibly the coordinator could handle leader changes as well? Not sure, to be honest. I guess I'm going into implementation details again.

One thing to think about is the resiliency of the coordinator. If we have a stateless service like Gateway as the coordinator, then how do we handle restarts of coordinator. How would it know about the ongoing backups started before its restart etc.. This is something that we can look at during the prototyping.

**Comment by deepthidevaki** [comment]
> * Having the StreamProcessor process the command to take a backup...as we're talking about an engine interface, is this something we still want to do? I assume it would be a stream processor we (our team) maintain, outside of the engine. Could this be a problem? Do we need to access the state or something like that? From reading into it, it looks like we'd need to fit this concept into the API the automation team would have to implement somewhat. Can we have a clearer separation of concern from our POV? i.e. we deal with logs, and whatever the snapshot store has (which as far as we're concerned is some metadata + a directory of files + SFV for checksum). Having the engine process these commands couples how we take backup with the engine implementation. Ideally we don't need that.

Here "process" doesn't mean the engine has to process the command. It could be considered as pre-process. After streamProcessor reads a record, it checks for the backup information in the record, triggers backup if needed and then send the record to the engine. 
Yes. We need to access the state to store ceckpointId and checkpointPosition. But the streamProcessor needs to access the state to store processedPositon anyway. The state should provide some api to store these "metadata" from the streamprocessor.
